var/spack/repos/builtin/packages/mvapich2/package.py

```diff
@@ -347,11 +348,40 @@ def setup_dependent_build_environment(self, env, dependent_spec):
         env.set('MPICH_F90', spack_fc)
         env.set('MPICH_FC', spack_fc)
 
-    def setup_compiler_environment(self, env):
-        # For Cray MPIs, the regular compiler wrappers *are* the MPI wrappers.
-        # Cray MPIs always have cray in the module name, e.g. "cray-mvapich"
+    def _decide_if_external_is_cray(self):
+        '''For Cray MPIs, the regular compiler wrappers *are* the MPI wrappers.
+           Cray MPIs always have cray in the module name, e.g. "cray-mvapich"
+           make no assumption about where the module may be in the module list.
+
+           inspect external modules and:
+               - return True if we believe it is Cray
+               - return False otherwise
+        '''
         external_modules = self.spec.external_modules
-        if external_modules and 'cray' in external_modules[0]:
+        expected_prefix = 'cray-mvapich'
+        have_cray = False
+        if external_modules:
+            tty.msg('Detected external modules, will test for a'
+                    ' module that starts with {0}'.format(expected_prefix))
+            for module_name in external_modules:
+                if module_name.startswith(expected_prefix):
+                    have_cray = True
+                    msg = 'Found module with name {0} - assuming Cray environment'
+                    tty.msg(msg.format(module_name))
+        return have_cray
```

Comment:
```suggestion
        expected_prefix = 'cray-mvapich'
        return any(m.startswith(expected_prefix) 
                          for m in external_modules)
```

Reply:
Will this work? is `external_modules` promised to be defined? If it's `None`, then the for loop will throw (I think)

Reply:
tentatively, I've keep my code organization. The primary value of the more verbose code is that info can be fed back to user (changed `tty.msg => tty.debug`.

IMO, that type of feedback is really important - particularly given the issues that seem to come up with modules.

Reply:
I've kept my code for now - it allows messages to get printed.

Moreover, You'd need to have a test for `external_modules` anyway - if it's none you choke. I suppose I may need a string test for `external_modules` or an assertion that `external_modules` is a list.

I'll add a `six` strings type test and repush.

Resolving this for now.